clientupdate/clientupdate.go

```diff
@@ -822,7 +825,179 @@ func (up *updater) updateFreeBSD() (err error) {
+		return err
+	}
+	if err := os.Remove(dlPath); err != nil {
+		up.Logf("failed to clean up %q: %w", dlPath, err)
+	}
+	if err := restartSystemdUnit(); err != nil {
+		if errors.Is(err, errors.ErrUnsupported) {
+			up.Logf("Tailscale binaries updated successfully.\nPlease restart tailscaled to finish the update.")
+		} else {
+			up.Logf("Tailscale binaries updated successfully, but failed to restart tailscaled: %s.\nPlease restart tailscaled to finish the update.", err)
+		}
+	} else {
+		up.Logf("Success")
+	}
+
+	return nil
+}
+
+func (up *updater) downloadLinuxTarball(ver string) (string, error) {
+	dlDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", err
+	}
+	dlDir = filepath.Join(dlDir, "tailscale-update")
+	if err := os.MkdirAll(dlDir, 0700); err != nil {
+		return "", err
+	}
+	pkgsPath := fmt.Sprintf("%s/tailscale_%s_%s.tgz", up.track, ver, runtime.GOARCH)
+	dlPath := filepath.Join(dlDir, path.Base(pkgsPath))
+	if err := up.downloadURLToFile(pkgsPath, dlPath); err != nil {
+		return "", err
+	}
+	return dlPath, nil
+}
+
+func (up *updater) unpackLinuxTarball(path string) error {
+	tailscale, tailscaled, err := binaryPaths()
+	if err != nil {
+		return err
+	}
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	gr, err := gzip.NewReader(f)
+	if err != nil {
+		return err
+	}
+	defer gr.Close()
+	tr := tar.NewReader(gr)
+	files := make(map[string]int)
+	wantFiles := map[string]int{
+		"tailscale":  1,
+		"tailscaled": 1,
+	}
+	for {
+		th, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed extracting %q: %w", path, err)
+		}
+		// TODO(awly): try to also extract tailscaled.service. The tricky part
+		// is fixing up binary paths in that file if they differ from where
+		// local tailscale/tailscaled are installed. Also, this may not be a
+		// systemd distro.
+		switch filepath.Base(th.Name) {
+		case "tailscale":
+			files["tailscale"]++
+			if err := writeFile(tr, tailscale+".new", 0755); err != nil {
+				return fmt.Errorf("failed extracting the new tailscale binary from %q: %w", path, err)
+			}
+		case "tailscaled":
+			files["tailscaled"]++
+			if err := writeFile(tr, tailscaled+".new", 0755); err != nil {
+				return fmt.Errorf("failed extracting the new tailscaled binary from %q: %w", path, err)
+			}
+		}
+	}
+	if !maps.Equal(files, wantFiles) {
+		return fmt.Errorf("%q has missing or duplicate files: got %v, want %v", path, files, wantFiles)
+	}
+
+	// Only place the files in final locations after everything extracted correctly.
+	if err := os.Rename(tailscale+".new", tailscale); err != nil {
+		return err
+	}
+	up.Logf("Updated %s", tailscale)
+	if err := os.Rename(tailscaled+".new", tailscaled); err != nil {
+		return err
+	}
+	up.Logf("Updated %s", tailscaled)
+	return nil
+}
+
+func writeFile(r io.Reader, path string, perm os.FileMode) error {
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, perm)
```

Comment:
Could use `O_WRONLY` instead of `O_RDWR`. Also `syscall.O_NOFOLLOW` might be a good idea for belt-and-suspenders paranoia? (Maybe not, but maybe.)

Reply:
Added `O_WRONLY`.
`O_NOFOLLOW` is not portable (i.e. doesn't compile on Windows), so I'd rather leave it out.

Reply:
I can't quite shake the fear of a symlink attack when writing as root. `O_NOFOLLOW` is present on macOS and FreeBSD as well as Linux, leaving only Windows, suggesting that a build tag approach wouldn't get too out of hand.

According to MSDN, https://learn.microsoft.com/en-us/windows/win32/fileio/symbolic-link-effects-on-file-systems-functions#createfile-and-createfiletransacted, `FILE_FLAG_OPEN_REPARSE_POINT` should work (in combination with `CREATE_ALWAYS` and/or `TRUNCATE_EXISTING`. All are present in the `syscall` package.

Alas, in my testing just now, passing `FILE_FLAG_OPEN_REPARSE_POINT | CREATE_ALWAYS | TRUNCATE_EXISTING` results in silent failure. `os.OpenFile` does not return an `error`. I think we'd have to use the actual `syscall.CreateFile` (which does exist, if we ever need to go that far).

Reply:
Would it be easier to just `os.Remove` the target first and then write a fresh file?

Reply:
Updated to do a delete before writing the file. In case there is a random symlink at the destination, it will just get replaced with a regular file instead of failing the update.
PTAL

Reply:
That would allow a local attacker who can place a symlink to cause root to remove any file, right? (Instead of overwriting any file with a tailscale binary.)

Alternate strategies:

* Tag this function as building only on Linux, and add a comment to the effect that if people want to build and use it on Windows, they should come talk to us about the bug
* Create a new directory and ensure that only root can write to it, thus ensuring that local attackers can't create symlinks in it (we may already be creating a new directory? if so adding the 1 check is easy)
* Document it as a TODO (possibly with a bug link) and come back to it later

Reply:
> That would allow a local attacker who can place a symlink to cause root to remove any file, right? (Instead of overwriting any file with a tailscale binary.)

I added a test specifically for this ;)
`os.Remove` unlinks a file and does not follow symlinks https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/os/file_unix.go;l=364